core/frontend/src/main/scala/io/udash/routing/RoutingEngine.scala

```diff
@@ -79,39 +80,39 @@ class RoutingEngine[S <: State : ClassTag : ImmutableValue](routingRegistry: Rou
     *
     * @param callback Callback getting StateChangeEvent as arguments
     */
-  def onStateChange(callback: StateChangeEvent[S] => Any): Registration = {
+  def onStateChange(callback: StateChangeEvent[HierarchyRoot] => Any): Registration = {
     callbacks += callback
     new Registration {
       override def cancel(): Unit = callbacks -= callback
     }
   }
 
   /** @return Current routing state */
-  def currentState: S = currentStateProp.get
+  def currentState: HierarchyRoot = currentStateProp.get
 
   /** @return Property reflecting current routing state */
-  def currentStateProperty: ReadableProperty[S] = currentStateProp.transform(identity)
+  def currentStateProperty: ReadableProperty[HierarchyRoot] = currentStateProp.transform(identity)
 
   @tailrec
-  private def getStatePath(forState: S, acc: List[S] = Nil): List[S] = forState match {
-    case state if state == rootState => rootState :: acc
-    case state: S => getStatePath(state.parentState.asInstanceOf[S], state :: acc)
+  private def getStatePath(forState: Option[HierarchyRoot], acc: List[HierarchyRoot] = Nil): List[HierarchyRoot] = forState match {
+    case Some(state) if state == rootState => rootState :: acc
```

Comment:
I think this branch is not necessary if `rootState.parentState` is always meant to return `None`.